lms/djangoapps/courseware/module_render.py

```diff
@@ -920,39 +926,37 @@ def handle_xblock_callback_noauth(request, course_id, usage_id, handler, suffix=
         return _invoke_xblock_handler(request, course_id, usage_id, handler, suffix, course=course)
 
 
-def handle_xblock_callback(request, course_id, usage_id, handler, suffix=None):
+class XblockCallbackView(APIView):
     """
-    Generic view for extensions. This is where AJAX calls go.
-
-    Arguments:
-        request (Request): Django request.
-        course_id (str): Course containing the block
-        usage_id (str)
-        handler (str)
-        suffix (str)
-
-    Raises:
-        Http404: If the course is not found in the modulestore.
+    Class-based view for extensions. This is where AJAX calls go.
+    Return 403 error if the user is not logged in. Raises Http404 if
+    the location and course_id do not identify a valid module, the module is
+    not accessible by the user, or the module raises NotFoundError. If the
+    module raises any other error, it will escape this function.
     """
-    # NOTE (CCB): Allow anonymous GET calls (e.g. for transcripts). Modifying this view is simpler than updating
-    # the XBlocks to use `handle_xblock_callback_noauth`...which is practically identical to this view.
-    if request.method != 'GET' and not request.user.is_authenticated():
-        return HttpResponseForbidden()
+    authentication_classes = (JwtAuthentication, SessionAuthentication, OAuth2Authentication,)
+    permission_classes = (IsAuthenticated,)
 
-    request.user.known = request.user.is_authenticated()
+    # This makes all operations atomic, and it's required to work nicely with the try/catch that can appear in tests.
+    # See https://docs.djangoproject.com/en/1.8/topics/db/transactions/#controlling-transactions-explicitly
+    @transaction.atomic
+    def dispatch(self, *args, **kwargs):
+        return super(XblockCallbackView, self).dispatch(*args, **kwargs)
```

Comment:
Is there any way we could move this into the failing tests instead? We try to avoid adding code for tests' sake only.

Reply:
It's strange but today I didn't get the error when running the tests without the `@atomic` part, so I suppose it must be something related to the test environment, devstack, or sqlite3.
I removed the `@atomic`.

For reference, this is the original error that made me need the `@atomic`:
```

self = <lms.djangoapps.courseware.tests.test_module_render.TestHandleXBlockCallback testMethod=test_completion_events_1___complete_____completion___0_625___0_625_>

    @wraps(func)
    def wrapper(self):
>       return func(self, *args, **kwargs)

../venvs/edxapp/local/lib/python2.7/site-packages/ddt.py:114: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../venvs/edxapp/local/lib/python2.7/site-packages/xblock/plugin.py:187: in _inner
    return func(*args, **kwargs)
lms/djangoapps/courseware/tests/test_module_render.py:708: in test_completion_events
    self.assertEqual(response.status_code, 200)
/usr/lib/python2.7/contextlib.py:35: in __exit__
    self.gen.throw(type, value, traceback)
openedx/core/djangoapps/waffle_utils/__init__.py:136: in override
    yield
/usr/lib/python2.7/contextlib.py:35: in __exit__
    self.gen.throw(type, value, traceback)
openedx/core/djangoapps/waffle_utils/__init__.py:162: in override_in_model
    yield
../venvs/edxapp/local/lib/python2.7/site-packages/waffle/testutils.py:72: in __exit__
    self.obj.delete()
../venvs/edxapp/local/lib/python2.7/site-packages/waffle/models.py:89: in delete
    ret = super(BaseModel, self).delete(*args, **kwargs)
../venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/base.py:896: in delete
    collector.delete()
../venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/deletion.py:309: in delete
    query.delete_batch(pk_list, self.using)
../venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/sql/subqueries.py:41: in delete_batch
    self.do_query(self.get_meta().db_table, self.where, using=using)
../venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/sql/subqueries.py:26: in do_query
    self.get_compiler(using).execute_sql(NO_RESULTS)
../venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py:840: in execute_sql
    cursor.execute(sql, params)
../venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/utils.py:59: in execute
    self.db.validate_no_broken_transaction()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <django.db.backends.sqlite3.base.DatabaseWrapper object at 0x7f7f88e0f910>

    def validate_no_broken_transaction(self):
        if self.needs_rollback:
            raise TransactionManagementError(
>               "An error occurred in the current transaction. You can't "
                "execute queries until the end of the 'atomic' block.")
E           TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
```